prover/src/epoch_service.rs

```diff
@@ -22,41 +21,41 @@ const WAIT_UNTIL_MIN_DURATION: Duration = Duration::from_secs(5);
 pub struct EpochService {
     /// A subdivision of an epoch (in minutes or seconds)
     epoch_slice_duration: Duration,
-    /// Current epoch and epoch slice
-    pub current_epoch: Arc<RwLock<(Epoch, EpochSlice)>>,
+    /// Sender to notify when an epoch / epoch slice has just changed
+    epoch_sender: Sender<(Epoch, EpochSlice)>,
+    /// Receiver that can be cloned for receivers
+    epoch_receiver: Receiver<(Epoch, EpochSlice)>,
     /// Genesis time (aka when the service has been started at the first time)
     genesis: DateTime<Utc>,
-    /// Channel to notify when an epoch / epoch slice has just changed
-    pub epoch_changes: Arc<Notify>,
 }
 
 impl EpochService {
+    pub fn epoch_changes(&self) -> Receiver<(Epoch, EpochSlice)> {
+        self.epoch_receiver.clone()
+    }
+
     pub(crate) async fn listen_for_new_epoch(&self) -> Result<(), AppError> {
         let epoch_slice_count =
             Self::compute_epoch_slice_count(EPOCH_DURATION, self.epoch_slice_duration);
         debug!("epoch slice in an epoch: {}", epoch_slice_count);
 
-        let (mut current_epoch, mut current_epoch_slice, mut wait_until) =
+        let (mut current_epoch, mut current_epoch_slice, mut wait_until) = loop {
             match self.compute_wait_until(&|| Utc::now(), &|| tokio::time::Instant::now()) {
-                Ok((current_epoch, current_epoch_slice, wait_until)) => {
-                    (current_epoch, current_epoch_slice, wait_until)
-                }
-                Err(_e) => {
-                    // sleep and try again (only one retry)
-                    tokio::time::sleep(WAIT_UNTIL_MIN_DURATION).await;
-                    self.compute_wait_until(&|| Utc::now(), &|| tokio::time::Instant::now())?
+                Ok(result) => break result,
+                Err(WaitUntilError::OutOfRange(_)) => {
+                    // If we're at a boundary, just wait a tiny bit and recalculate
+                    tokio::time::sleep(Duration::from_millis(10)).await;
```

Comment:
Any reason about using 10ms here? It seems quite low? Can you define a constant value here?